src/frontend/src/optimizer/rule/filter_join.rs

```diff
@@ -104,6 +104,15 @@ impl FilterJoinRule {
         Box::new(FilterJoinRule {})
     }
 
+    fn and(left: Option<Condition>, right: Option<Condition>) -> Option<Condition> {
```

Comment:
It seems these 13 lines are enough to fix the issue, and the other 171 lines are doing other things. So I highly suggest splitting them this time, which is not very hard - essentially if you read on.

Talking about this 13-line fix, is it better to avoid this confusing `Option<Condition>` at all? Given an empty condition / the absent of a condition means `true`, and we do have `Condition::true_cond()` and `Condition::always_true()`, shall we just use them instead? Btw there is also `Condition::and`. @xxchan correct me if I misunderstood `Option<Condition>` vs `Condition::true_cond()`.

If the better fix by avoiding `Option<Condition>` is correct, it can be a new PR titled 'fix ...', and this PR can be retitled 'add FilterMerge and FilterRemove'.

Reply:
It makes sense to remove options